src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1040 @@
+--
+-- Only use for integrals size less than or equal to that of 'Word32'.
+unbiasedWordMult32RM :: (StatefulGen g m, Integral a) => (a, a) -> g -> m a
+unbiasedWordMult32RM (b, t) g
+  | b <= t    = (+b) . fromIntegral <$> unbiasedWordMult32 (fromIntegral (t - b)) g
+  | otherwise = (+t) . fromIntegral <$> unbiasedWordMult32 (fromIntegral (b - t)) g
+{-# SPECIALIZE unbiasedWordMult32RM :: StatefulGen g m => (Word8, Word8) -> g -> m Word8 #-}
+
+-- | Uniformly generate Word32 in @[0, s]@.
+unbiasedWordMult32 :: StatefulGen g m => Word32 -> g -> m Word32
+unbiasedWordMult32 s g
+  | s == maxBound = uniformWord32 g
+  | otherwise = unbiasedWordMult32Exclusive (s+1) g
+{-# INLINE unbiasedWordMult32 #-}
+
+-- | See [Lemire's paper](https://arxiv.org/pdf/1805.10941.pdf),
+-- [O\'Neill's
+-- blogpost](https://www.pcg-random.org/posts/bounded-rands.html) and
+-- more directly [O\'Neill's github
+-- repo](https://github.com/imneme/bounded-rands/blob/3d71f53c975b1e5b29f2f3b05a74e26dab9c3d84/bounded32.cpp#L234).
+-- N.B. The range is [0,t) **not** [0,t].
+unbiasedWordMult32Exclusive :: forall g m . StatefulGen g m => Word32 -> g -> m Word32
+unbiasedWordMult32Exclusive r g = go
+  where
+    t :: Word32
+    t = (-r) `mod` r -- Calculates 2^32 `mod` r!!!
+    go :: StatefulGen g m => m Word32
+    go = do
+      x <- uniformWord32 g
+      let m :: Word64
+          m = fromIntegral x * fromIntegral r
+          l :: Word32
+          l = fromIntegral m
+      if l >= t then return (fromIntegral $ m `shiftR` 32) else go
+
+-- | This only works for unsigned integrals
+unsignedBitmaskWithRejectionRM ::
+     (StatefulGen g m, FiniteBits a, Num a, Ord a, Uniform a)
+  => (a, a)
+  -> g
+  -> m a
+unsignedBitmaskWithRejectionRM (bottom, top) gen
+  | bottom == top = pure top
+  | otherwise = (b +) <$> unsignedBitmaskWithRejectionM uniformM r gen
+  where
+    (b, r) = if bottom > top then (top, bottom - top) else (bottom, top - bottom)
+{-# INLINE unsignedBitmaskWithRejectionRM #-}
+
+-- | This works for signed integrals by explicit conversion to unsigned and abusing overflow
+signedBitmaskWithRejectionRM ::
+     (Num a, Num b, Ord b, Ord a, FiniteBits a, StatefulGen g f, Uniform a)
+  => (b -> a)
+  -> (a -> b)
+  -> (b, b)
+  -> g
+  -> f b
+signedBitmaskWithRejectionRM toUnsigned fromUnsigned (bottom, top) gen
+  | bottom == top = pure top
+  | otherwise =
+    (b +) . fromUnsigned <$> unsignedBitmaskWithRejectionM uniformM r gen
+    -- This works in all cases, see Appendix 1 at the end of the file.
+  where
+    (b, r) =
+      if bottom > top
+        then (top, toUnsigned bottom - toUnsigned top)
+        else (bottom, toUnsigned top - toUnsigned bottom)
+{-# INLINE signedBitmaskWithRejectionRM #-}
+
+unsignedBitmaskWithRejectionM ::
+  forall a g m . (Ord a, FiniteBits a, Num a, StatefulGen g m) => (g -> m a) -> a -> g -> m a
+unsignedBitmaskWithRejectionM genUniformM range gen = go
```

Comment:
A comment about semantics of `unsignedBitmaskWithRejectionM` and `signedBitmaskWithRejectionRM` would be helpful. What properties should `toUnsigned` / `fromUnsigned` satisfy?

Reply:
I added some comments and a link to the post that describes bitmask with rejection approach.